> #### 5조

# 1번 (25점)

#### 문제 1-1. 각 분류 모델의 주요 특징을 간략하게 설명해주세요.  (5점, 각 1점)

1. GBM
: 여러 개의 약한 예측 모델들을 결합하여 강력한 예측 모델을 만드는 기법

2. XGB (XGBoost)
:과적합을 방지하기 위한 정규화를 포함한 고성능 그래디언트 부스팅 라이브러리

3. LGBM (LightGBM)
: 그래디언트 부스팅 프레임워크로, 리프 중심 트리 분할(Leaf-wise tree growth) 방식을 사용하여 모델의 정확도를 높이면서도 계산 비용을 낮춤.

4. SVM
: 데이터를 분류하기 위한 경계선(결정 경계)을 정의하는 모델로, 분류를 위한 최적의 경계선을 찾는 것을 목표로 함.

5. KNN
:: KNN은 분류하고자 하는 샘플에 가장 가까운 K개의 이웃 샘플들의 정보를 바탕으로 해당 샘플의 분류나 회귀 결과를 예측하는 비모수 방식의 알고리즘

### **1-2.  GBM의 가장 큰 단점은 무엇인가요? (3점)**

a) 과적합 가능성이 낮음

b) 속도가 매우 빠름

c) 하이퍼파라미터 튜닝이 쉬움

d) 학습 시간이 오래 걸릴 수 있음

In [1]:
# 정답 : d

### **1-3. XGB의 주요 특징 중 하나는 무엇인가요? (3점)**

a) 병렬 처리 기능을 갖추고 있지 않음

b) 속도가 느리고 메모리를 많이 사용함

c) 규제 기능을 갖추어 과적합을 방지함

d) 하이퍼파라미터 튜닝이 필요하지 않음

In [2]:
# 정답 : c

### **1-4. LGBM의 주요 특징 중 하나는 무엇인가요? (3점)**

a) 학습 시간이 오래 걸림

b) 고차원 데이터에 적합하지 않음

c) 대규모 데이터셋에 효율적으로 적용됨

d) 규제 기능이 없음

In [3]:
# 정답 : c

### **1-5. SVM은 어떤 종류의 학습 알고리즘인가요? (3점)**

a) 지도 학습
b) 비지도 학습
c) 준지도 학습
d) 강화 학습

In [6]:
# 정답 : a

### **1-6. K-NN 알고리즘에서 k값이 너무 작을 경우 발생할 수 있는 문제는 무엇인가요? (3점)**

a) 과적합될 가능성이 높음
b) 모델이 너무 단순해짐
c) 노이즈의 영향을 줄이기 어려움
d) 모델이 너무 복잡해짐

In [5]:
# 정답 : a

### **1-7. XGBoost 모델에서 가장 중요한 하이퍼파라미터는 무엇이고, 이를 조정함으로써 어떤 효과를 기대할 수 있나요? (5점)**

In [7]:
# learning_rate : 각 단계에서 가중치 업데이트의 크기를 결정

---

# 2번

### 기본세팅

In [8]:
# 필요한 라이브러리
import numpy as np
import pandas as pd

import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [9]:
# 데이터 로드
iris = load_iris()
X, y = iris.data, iris.target

# 학습 데이터와 테스트 데이터로 분리
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### 2-1. GBM 예제. 다음 코드를 완성해주세요. (5점)

In [11]:
from sklearn.ensemble import GradientBoostingClassifier

gbm_model = GradientBoostingClassifier()
gbm_model.fit(X_train, y_train)
gbm_pred = gbm_model.predict(X_test)
print("GBM Accuracy:", accuracy_score(y_test, gbm_pred))

GBM Accuracy: 1.0


### 2-2. XGB 예제. 다음 코드를 완성해주세요. (5점)

##### XGBoost 설치하기
: 아나콘다 Command 창을 연뒤에 ‘conda install -c anaconda py-xgboost’ 명령어 입력
안되면 구글 코랩을 사용해주세요!

In [12]:
import xgboost as xgb
xgb_model =  xgb.XGBClassifier()
xgb_model.fit(X_train, y_train)
xgb_pred = xgb_model.predict(X_test)
print("XGBoost Accuracy:", accuracy_score(y_test, xgb_pred))

XGBoost Accuracy: 1.0


### 2-3. LGBM 예제. 다음 코드를 완성해주세요.(5점)

In [13]:
import lightgbm as lgb
lgb_model =  lgb.LGBMClassifier()
lgb_model.fit(X_train, y_train)
lgb_pred = lgb_model.predict(X_test)
print("LightGBM Accuracy:", accuracy_score(y_test, lgb_pred))

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000158 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 91
[LightGBM] [Info] Number of data points in the train set: 120, number of used features: 4
[LightGBM] [Info] Start training from score -1.098612
[LightGBM] [Info] Start training from score -1.073920
[LightGBM] [Info] Start training from score -1.123930
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No furth

### 2-4. SVM 예제. 다음 코드를 완성해주세요. (5점)

In [14]:
from sklearn.svm import SVC
svm_model = SVC()
svm_model.fit(X_train, y_train)
svm_pred = svm_model.predict(X_test)
print("SVM Accuracy:", accuracy_score(y_test, svm_pred))

SVM Accuracy: 1.0


### 2-5. K-NN 예제. 다음 코드를 완성해주세요. (5점)

In [15]:
from sklearn.neighbors import KNeighborsClassifier
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)
print("K-NN Accuracy:", accuracy_score(y_test, knn_pred))

K-NN Accuracy: 1.0


---

---

> #### 1조

## 1번 (24점)

#### 문제 1-1. 과적합과 과소적합이 발생하는 원인을 각각 한 줄로 서술하시오. (3점)

답:

과적합: 모델이 학습 데이터에 너무 잘 맞추어져 있어 새로운 데이터에 대한 일반화 능력이 떨어지는 상황입니다.

과소적합: 모델이 너무 단순하여 학습 데이터의 패턴을 충분히 학습하지 못하고, 새로운 데이터에 대해서도 예측 성능이 낮은 상황입니다.

#### 문제 1-2. 과적합을 대처하는 방법으로 옳지 않은 것을 고르시오. (3점)

1. 데이터 양 늘리기

2. Feature의 개수를 줄이기

3. Parameter가 더 많은 복잡한 모델을 선택하기

4. 훈련데이터의 잡음을 줄이기

5. 정규화 기법 사용하기

답: 3

#### 문제 1-3. 정확도의 개념과 한계에 대해 적어주세요. (3점)

정확도(accuracy)는 전체 데이터 중 모델이 올바르게 예측한 데이터의 비율을 의미
<br>
<br>
한계: 정확도는 불균형한 데이터셋에서 모델의 성능을 과대평가하거나 과소평가할 수 있음.

- 오차 행렬은 다음과 같습니다.
![image.png](attachment:7cf62483-a6ae-4800-99c7-7e47a01e9197.png)

#### 문제 1-4. TP, FP, FN, TN에 대해 간략히 설명해주세요. (3점)

TP (True Positive): 실제 양성을 양성으로 정확하게 예측한 경우의 수
<br>
FP (False Positive): 실제 음성을 양성으로 잘못 예측한 경우의 수 (1종 오류)
<br>
FN (False Negative): 실제 양성을 음성으로 잘못 예측한 경우의 수 (2종 오류)
<br>
TN (True Negative): 실제 음성을 음성으로 정확하게 예측한 경우의 수

#### 문제 1-5. 위와 같은 오차 행렬에서 정확도, 정밀도, 재현율은 어떻게 정의될까요? (TN, FP, FN, TP를 사용하여 수식으로 표현해주세요. ) (3점)

답 :
1) 정확도 : (TP+TN)/(TP+TN+FP+FN)
<br>
2) 정밀도 : TP / (TP+FP)
<br>
3) 재현율 : TP / (TP+FN)

#### 문제 1-6. 정밀도와 재현율을 결합하여 F1스코어의 수식을 적어주세요. (3점)

답 : F1 = 2 x (정밀도+재현율) / 정밀도x 재현율
​


#### 문제 1-7. FPR과 TPR의 수식을 적고 각각이 의미하는 바를 설명해주세요. (3점)

답 : FPR = FP/(FP+TN)

답 : TPR = TP/(TP+FN)

#### 문제 1-8. AUC 스코어를 ROC 곡선을 이용하여 설명해주세요. (3점)

답 : AUC (Area Under the Curve) 스코어는 ROC (Receiver Operating Characteristic) 곡선 아래의 면적을 의미합니다. ROC 곡선은 다양한 분류 임계값에서의 TPR(재현율)과 FPR(거짓 양성 비율)을 플로팅한 것입니다. AUC 스코어는 0과 1 사이의 값을 가지며, 값이 클수록 모델의 성능이 더 좋다고 평가할 수 있습니다

## 2번 (26점)

(문제가 아닙니다. 코드를 실행하세요.)

In [16]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.linear_model import Ridge, Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

# seaborn을 통해 타이타닉 데이터셋 로드
titanic = sns.load_dataset('titanic')

# 데이터 전처리: 필요한 특성 선택 및 결측치 처리
titanic = titanic[['survived', 'pclass', 'age', 'sibsp', 'parch', 'fare']].dropna()

X = titanic.drop('survived', axis=1)
y = titanic['survived']

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 특성 스케일링
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


#### 문제 2-1. 릿지 회귀를 이용하여 과적합을 방지하는 코드를 완성하시오. (4점)

In [19]:
# 릿지 회귀 모델
ridge_model = Ridge(alpha=1.0)
ridge_model.fit(X_train_scaled, y_train)
ridge_pred = ridge_model.predict(X_test_scaled)
ridge_mse = mean_squared_error(y_test, ridge_pred)
print(f'Ridge Regression MSE: {ridge_mse}')

Ridge Regression MSE: 0.19864875882344685


올바른 출력값: Ridge Regression MSE: 0.19864875882344687

#### 문제 2-2. 라쏘 회귀를 이용하여 과적합을 방지하는 코드를 완성하시오. (4점)

In [20]:
# 라쏘 회귀 모델
lasso_model = Lasso(alpha=0.1)
lasso_model.fit(X_train_scaled, y_train)
lasso_pred = lasso_model.predict(X_test_scaled)
lasso_mse = mean_squared_error(y_test, lasso_pred)
print(f'Lasso Regression MSE: {lasso_mse}')

Lasso Regression MSE: 0.2138317931536248


올바른 출력값: Lasso Regression MSE: 0.2138317931536248

(문제가 아닙니다. 다음 코드를 실행하세요.)

In [17]:
from sklearn.base import BaseEstimator
import numpy as np

class MyDummyClassifier(BaseEstimator):
    # fit() 메서드는 아무것도 학습하지 않음.
    def fit(self, X, y=None):
        pass
    # predict() 메서드는 단순히 Sex피처가 1이면 0, 그렇지 않으면 1로 예측함
    def predict(self, X):
        pred = np.zeros((X.shape[0],1)) # 0으로 구성된 다차원 array 생성
        for i in range(X.shape[0]):
            if X['Sex'].iloc[i] == 1 : # 위치기반 인덱싱
                pred[i] = 0
            else:
                pred[i] = 1

        return pred

from sklearn.preprocessing import LabelEncoder
# Null 처리 함수
def fillna(df):
    df['Age'].fillna(df['Age'].mean(), inplace=True)
    df['Cabin'].fillna('N', inplace=True)
    df['Embarked'].fillna('N', inplace=True)
    df['Fare'].fillna(0, inplace=True)
    return df
# 머신러닝 알고리즘에 불필요한 피처 제거
def drop_features(df):
    df.drop(['PassengerId', 'Name', 'Ticket'], axis=1, inplace=True)
    return df

# 레이블 인코딩 수행
def format_features(df):
    df['Cabin'] = df['Cabin'].str[:1]
    features = ['Cabin', 'Sex', 'Embarked']
    for feature in features:
        le = LabelEncoder()
        le = le.fit(df[feature])
        df[feature] = le.transform(df[feature])
    return df
# 앞에서 설정한 데이터 전처리 함수 호출
def transform_features(df):
    df = fillna(df)
    df = drop_features(df)
    df = format_features(df)
    return df


In [21]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# 타이타닉 데이터셋 로드
titanic_df = pd.read_csv('/content/titanic_train.csv')

y_titanic_df = titanic_df['Survived']
X_titanic_df = titanic_df.drop('Survived', axis=1)

X_titanic_df = transform_features(X_titanic_df)
X_train,X_test, y_train, y_test = train_test_split(X_titanic_df, y_titanic_df,
                                                   test_size=0.2, random_state=0)


#### 문제 2-3. 코드의 '???' 부분을 채워서 실행해주세요. (5점)

In [23]:
#위에서 생성한 DummyClassifier를 이용해 학습/예측/평가 수행
myclf = MyDummyClassifier()
myclf.fit(X_train, y_train)

mypredictions = myclf.predict(X_test)
print('Dummy Classifier의 정확도는: {0:.4f}'.format(accuracy_score(y_test, mypredictions)))

Dummy Classifier의 정확도는: 0.7877


[output]
![image.png](attachment:e36ed738-3ef1-4046-8095-3be0d9322e63.png)

#### 문제 2-4. 코드의 '???' 부분을 채워서 실행해주세요. (5점)

In [24]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

def get_clf_eval(y_test, pred):
    confusion = confusion_matrix(y_test, pred)
    accuracy = accuracy_score(y_test, pred)
    precision = precision_score(y_test, pred)
    recall = recall_score(y_test, pred)
    print('오차 행렬')
    print(confusion)
    print('정확도: {0:.4f}, 정밀도: {1:.4f}, 재현율: {2:.4f}'.format(accuracy, precision, recall))

In [25]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

#원본 데이터를 재로딩, 데이터 가공, 학습 데이터/테스트 데이터 분할
titanic_df = pd.read_csv('titanic_train.csv')
y_titanic_df = titanic_df['Survived']
X_titanic_df = titanic_df.drop('Survived', axis=1)
x_titanic_df = transform_features(X_titanic_df)

X_train,X_test,y_train,y_test = train_test_split(X_titanic_df, y_titanic_df,
                                                test_size=0.2, random_state=11)

Ir_clf = LogisticRegression(solver='liblinear')

Ir_clf.fit(X_train,y_train)
pred = Ir_clf.predict(X_test)
get_clf_eval(y_test, pred)

오차 행렬
[[108  10]
 [ 14  47]]
정확도: 0.8659, 정밀도: 0.8246, 재현율: 0.7705


[output]
![image.png](attachment:13482f4b-8f83-4628-952d-fa1727156eee.png)705

#### 문제 2-5. F1 스코어를 구하기 위한 아래 코드의 '???' 부분을 채워서 실행해주세요. (4점)

In [26]:
from sklearn.metrics import f1_score

f1 = f1_score(y_test, pred)
print('F1 스코어: {0:4f}'.format(f1))

F1 스코어: 0.796610


올바른 출력값)  F1 스코어: 0.796610

#### 문제 2-6. AOC 스코어를 구하기 위한 아래 코드의 '???' 부분을 채워서 실행해주세요. (4점)

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
lr_clf = LogisticRegression(solver='liblinear')
lr_clf.fit(X_train,y_train)
pred_proba = lr_clf.predict_proba(X_test)[:, 1]
roc_score = roc_auc_score(y_test, pred_proba)
print('ROC AUC 값: {0:.4f}'.format(roc_score))

ROC AUC 값: 0.8987


올바른 출력값)  ROC AUC 값: 0.8987